src/services/completions.ts

```diff
@@ -1346,7 +1348,7 @@ function createCompletionEntry(
         }
     }
 
-    if (origin?.kind === SymbolOriginInfoKind.TypeOnlyAlias) {
+    if ((origin?.kind === SymbolOriginInfoKind.TypeOnlyAlias) || (contextToken && contextToken.kind !== SyntaxKind.OpenBraceToken && completionKind === CompletionKind.ObjectPropertyDeclaration && preferences.includeCompletionsWithInsertText)) {

```

Comment:
We shouldn’t be setting `hasAction` unless we’re _definitely_ going to produce a code action. It looks like you’re setting `hasAction` for all object literal member completions (except the first member in the object), and then later just not returning code actions if a comma insertion isn’t actually needed. Instead, we should be checking _here_ if a comma insertion is needed, and only setting `hasAction` if it is.

Also, you should make a new `CompletionSource` to represent this check. That way, in `getCompletionEntryCodeActionsAndSourceDisplay`, rather than repeating the syntax/text-based check to see if we need a comma insertion, you can just let the `source` option tell you that’s what we’re doing.